rfcs/unboxed-types.md

```diff
@@ -0,0 +1,458 @@
+lists. Note that the existing `+.`, etc. primitives won't work on
+`#float`, because they have the wrong type. New ones for `#float` will
+be made available. (Same for `#int32`, etc.)
+
+Unboxed types will be allowed in record fields:
+
+    type t = {
+      foo : int;
+      bar : #float;
+      ...
+    }
+
+Because of how the OCaml runtime works, there are some restrictions on
+how boxed and unboxed types can be mixed in a record. If there are any
+fields which might contain pointers (that is, are of layout `value`
+but are not `immediate`), then all fields must be of layout
+`value`. So, you can mix `int` and `#float` in a single record, but
+not `string` and `#float`.
+
+## Unboxed tuples and records
+
+The second part of this proposal is to allow user-defined unboxed
+types (by contrast to what's above: user-defined boxed types
+containing unboxed fields). The simplest of these is unboxed tuples:
+
+    #( int * float * string )
+
+This is an unboxed tuple, and is passed around as three separate
+values.  Passing it to a function requires three registers.
+The layout of an unboxed tuple is the product of the layout of its
+components. In particular, the layout of `#( string * string )` is
+`value * value`, not `value`, and so it cannot be stored in, say, a
+list.
+
+Unboxed records are in theory just as straightforward, but have some
+extra complexity just to handle the various ways that types are
+defined and used. Suppose you write:
+
+    type t = { x : int32; y : int32 }
+
+Sometimes we want to refer to `t` as a boxed type, to store it in
+lists and other standard data structures. Sometimes we want to refer
+to `t` as an unboxed type, to avoid extra allocations (in particular,
+to inline it into other records). The language never infers which is
+which (boxing and allocation is always explicit in this design), so we
+need different syntax for both.  So, `type t = { ... }` defines *two*
+types, the boxed record `t : value` and the unboxed record `#t :
+int32 * int32`. They have different syntax for introduction and
+elimination:
+
+    let p : t = { x = a; y = b }
+    let px = p.x
+    let q : #t = #{ x = a; y = b }
+    let qx = q.#x
+
+An unboxed record like `q` cannot itself contain mutable
+fields. However, a boxed record can contain a mutable field of type
+`q`. This can be used to inline one record inside another, e.g.:
+
+    type s = { foo : int; mutable bar : #t }
+
+This defines a boxed record, layed out as three fields. We will
+support efficient updates of sub-parts of `bar`, at least with the
+following syntax:
+
+    x.bar <- { x.bar with y = y' }
+
+It would be nice to also allow the syntax `x.bar.#y <- y'`, although
+this is a little trickier to understand and implement, since it is
+`x.bar` rather than `x.bar.y` which is the mutable field.
+
+Type aliases do not automatically bind an unboxed variant, because
+the scoping of type aliases should not depend on whether they expand
+to records or not. So, if you write:
```

Comment:
@Drup says
I do not understand this. # is pretty much an operator on types that represent their unboxed representation, if available. Why shouldn't it be allowed to follow type aliases ?

@lpw25 says
The key here is that # is not actually an operator on types -- #foo is just the name that we give to an automatically created type, as far as the type system is concerned it has no actual relationship to foo. So #foo is a fully fledged member of its containing module -- a separate entity from foo.

Whether or not a type aliases to a record is not stable under substitution -- it can change if you e.g. apply a functor -- whereas you really want the items in a signature to be stable under substitution.

@Drup says
That doesn't really explain why # is not an operator, though.

If we use the proposed syntax type t : value with #t = ..., then there is no reason to make #t a name, we can deduce everything we need from the environment.

@lpw25 says
Well a proper type operator needs to make sense on all types (of the appropriate kind). Whereas this # is only well defined on some datatypes.

Now you could make a box operator that would work on anything with an appropriate kind, and indeed I think that would be good to add later. But you probably still end up wanting to provide access to the unboxed version of all existing record definitions -- so you'd probably still want the # pseudo operator such that:

```
type foo = { bar : int }
```
would secretly be something like:

```type #foo = #{ bar : int }
type foo = #foo box = { bar : int }
```

@lpw25 says
It might be worth noting that this # is actually pretty well aligned with how # types behave for classes.